ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -2450,6 +2453,233 @@ prop_connection_manager_counters serverAcc (ArbDataFlow dataFlow)
+
+             _ -> error ("Unexpected invalid transition: " ++ show (st, tt))
+
+           OutboundDupSt Ticking        -> case to of
+             -- Should preserve the timeout
+             InboundIdleSt Duplex -> verifyTimeouts st xs
+             OutboundDupSt Ticking -> verifyTimeouts st xs
+
+             -- Timeout terminating states
+             OutboundDupSt Expired ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+             DuplexSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+             InboundSt Duplex ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+             TerminatedSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             -- This state terminates the current timeout
+             -- and starts a new one
+             TerminatingSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts newState xs
+
+             _ -> error ("Unexpected invalid transition: " ++ show (st, tt))
+
+           OutboundIdleSt _             -> case to of
+             -- Timeout terminating states
+             InboundSt Duplex ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+             TerminatedSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             -- This state terminates the current timeout
+             -- and starts a new one
+             TerminatingSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= outboundTimeout
+               .&&. verifyTimeouts newState xs
+
+             _ -> error ("Unexpected invalid transition: " ++ show (st, tt))
+
+           TerminatingSt                -> case to of
+             -- Timeout terminating states
+             UnnegotiatedSt Inbound ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= timeWaitTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             TerminatedSt ->
+               counterexample (errorMsg tt t' t idleTimeout)
+               $ diffTime t t' <= timeWaitTimeout
+               .&&. verifyTimeouts Nothing xs
+
+             _ -> error ("Unexpected invalid transition: " ++ show (st, tt))
+
+           _ -> error ("Should be a \tau state: " ++ show st)
```

Comment:
> We actually have one more timeout which is not reflected on the state machine: handshake one. It starts in UnnegotiatedState _ and terminates on its first transition (to Outbound Unidirectional, InboundIdleState^\tau Duplex, InboundIdleState\tau Unidirectional or TerminatedState).

Oh yeah I could add this ! Thanks :) 

> Have you thought how to show some statistics which timeouts are we actually testing?

Like showing some sort of coverage? I know we have the state coverage for the outbound governor so one can infer some statistics from there, since we test a timeout everytime we reach a \tau state. If that coverage test shows good coverage for the \tau states then I think we are good

Reply:
Addressed